src/transpilation/resolve.ts

```diff
@@ -34,206 +36,184 @@ class ResolutionContext {
+        // If the import is not relative, resolve it relative to options.baseUrl if it is set
+        const relativeTo = isRelative ? fileDirectory : this.options.baseUrl ?? fileDirectory;
 
-    const possibleProjectFiles = [
-        resolvedPath, // JSON files need their extension as part of the import path, caught by this branch,
-        resolvedPath + ".ts", // Regular ts file
-        path.join(resolvedPath, "index.ts"), // Index ts file,
-        resolvedPath + ".tsx", // tsx file
-        path.join(resolvedPath, "index.tsx"), // tsx index
-    ];
+        // Check if file is a file in the project
+        const resolvedPath = path.join(relativeTo, dependency);
+        const fileFromPath = this.getFileFromPath(resolvedPath);
+        if (fileFromPath) return fileFromPath;
 
-    for (const possibleFile of possibleProjectFiles) {
-        if (isProjectFile(possibleFile, program)) {
-            return possibleFile;
+        // Check if this is a sibling of a required lua file
+        if (requiringFile.fileName.endsWith(".lua")) {
+            const luaFilePath = resolveLuaPath(fileDirectory, dependency, this.emitHost);
+            if (luaFilePath) {
+                return luaFilePath;
+            }
         }
-    }
-
-    // Check if this is a lua file in the project sources
-    const possibleLuaProjectFiles = [
-        resolvedPath + ".lua", // lua file in sources
-        path.join(resolvedPath, "index.lua"), // lua index file in sources
-        path.join(resolvedPath, "init.lua"), // lua looks for <require>/init.lua if it cannot find <require>.lua
-    ];
 
-    for (const possibleFile of possibleLuaProjectFiles) {
-        if (emitHost.fileExists(possibleFile)) {
-            return possibleFile;
+        // Not a TS file in our project sources, use resolver to check if we can find dependency
+        try {
+            const resolveResult = resolver.resolveSync({}, fileDirectory, dependency);
+            if (resolveResult) return resolveResult;
+        } catch (e) {
+            // resolveSync errors if it fails to resolve
         }
+
+        return undefined;
     }
 
-    // Check if this is a sibling of a required lua file
-    if (requiringFile.fileName.endsWith(".lua")) {
-        const luaFilePath = resolveLuaPath(fileDirectory, dependency, emitHost);
-        if (luaFilePath) {
-            return luaFilePath;
+    // value is false if already searched but not found
+    private pathToFile = new Map<string, string | false>();
+
+    private getFileFromPath(resolvedPath: string): string | undefined {
+        const existingFile = this.pathToFile.get(resolvedPath);
+        if (existingFile) return existingFile;
+        if (existingFile === false) return undefined;
+
+        const file = searchForFile(this);
+        this.pathToFile.set(resolvedPath, file ?? false);
+        return file;
+
+        function searchForFile(context: ResolutionContext) {
```

Comment:
I don't really get the point of making this a local function that is called just once, why not just keep this function code inlined like it was. Also why do you pass in `this` if you could just capture from context?

Reply:
The function was helpful because you can `return` out of it early. I moved it to a member function.